docs/IR.md

```diff
@@ -1,127 +1,372 @@
-|Complex Types|__complex64,complex128__|A complex number with either 32- or 64-bit real and imaginary parts.|
-|Other|__string__|Strings represent textual data. All strings are encoded using UTF-8.|
-|Other|__bool__|Boolean value represent data with only two values, typically _true_ and _false_.|
-|Other|__handle__|Handles are opaque types holding a 64-bit integer.|
-|Collections|__sparse and dense tensor__|Tensors are a generalization of vectors and matrices; whereas vectors have one dimension, and matrices two, tensors can have any number of dimensions, including zero. A zero-dimensional tensor is equivalent to a scalar.|
+|Variant | Type | Description | 
+|---|---|---|
+ONNX|dense tensors|Tensors are a generalization of vectors and matrices; whereas vectors have one dimension, and matrices two, tensors can have any number of dimensions, including zero. A zero-dimensional tensor is logically equivalent to a scalar value.
+ONNX-ML|sequence|Sequences represent dense, ordered, collections of elements that are of homogeneous types.
+ONNX-ML|map|Maps represent associative tables, defined by a key type and a value type.
 
-Extensibility
--------------
+ONNX currently does not define a sparse tensor type.
 
-ONNX is expected to evolve over time and provides features that enable users to experiment and implement additions before they are added to the specifications.
+#### Tensor shapes
 
-### Metadata
+In addition to element type, tensors have shape. A tensor shape is a list of records that define whether the tensor is a vector, a matrix, or a higher-dimensional value. For example, a 100x100 matrix has the shape [100,100].
 
-A model allows named metadata strings to be added via its `metadata_props` field, typically for use by tools such as converters, indexers, or documentation generators. Names are not prescribed, but some name recommendations are provided for implementations that want to record such concepts.
+The shape record is defined by 'TensorShapeProto':
 
-- `author`: the name of the person(s) who authored the model
-- `company`: the name of the company or organization that authored the model
-- `converted_from`: if converted from a different format, the name of the source format or framework
-- `license`: a human-readable name for a license, if applicable
-- `license_url`: the URL where the license text is provided
+```
+message TensorShapeProto {
+  message Dimension {
+    oneof value {
+      int64 dim_value = 1;
+      string dim_param = 2;
+    };
+  };
+  repeated Dimension dim = 1;
+}
+```
+Which is referenced by the Tensor type message:
+
+```
+  message Tensor {
+    optional TensorProto.DataType elem_type = 1;
+    optional TensorShapeProto shape = 2;
+  }
+```
+
+The empty list of dimension sizes, [], is a valid tensor shape, denoting a zero-dimension (scalar) value. A zero-dimension tensor is distinct from a tensor of unknown dimensionality, which is indicated by an absent 'shape' property in the Tensor record.
```

Comment:
Would you want to add a note here to explain that symbolic dimension variable is not unknown but variable-length? It's kind of confusing to my tiny brain.